lib/pbench/cli/agent/commands/tools/kill.py

```diff
@@ -0,0 +1,247 @@
+            pid = (tm_dir / self.file_name).read_text()
+        except FileNotFoundError:
+            return False
+        try:
+            self.procs_by_uuid[uuid] = psutil.Process(pid)
+        except psutil.NoSuchProcess:
+            return False
+        self.uuid_to_tmdir[uuid] = tm_dir
+        return True
+
+    def killem(self, echo: Callable[[str], None]) -> None:
+        """Kill all PIDs found, and their children."""
+        if not self.procs_by_uuid:
+            return
+        echo(f"Killing {self.display_name} PIDs ...")
+        # Clear out the stored data ahead of the killings.
+        procs_by_uuid, self.procs_by_uuid = self.procs_by_uuid, {}
+        uuid_to_tmdir, self.uuid_to_tmdir = self.uuid_to_tmdir, {}
+        for uuid, proc in procs_by_uuid.items():
+            pid = proc.pid
+            echo(f"\tKilling {pid} (from {uuid_to_tmdir[uuid]})")
+            try:
+                kill_family(proc)
+            except Exception as exc:
+                echo(f"\t\terror killing {pid}: {exc}", err=True)
+
+
+def gen_run_directories(pbench_run: pathlib.Path) -> Iterable[Tuple[pathlib.Path, str]]:
+    """Generate the list of run directories available under ${pbench_run},
+    yielding a Path object for that directory, along with its recorded
+    UUID.
+
+    Yields a tuple of the run directory Path object and associated UUID.
+    """
+    for entry in pbench_run.iterdir():
+        if not entry.is_dir():
+            continue
+        tm_dir = entry / "tm"
+        try:
+            uuid = (tm_dir / ".uuid").read_text()
+        except FileNotFoundError:
+            # This is either not a pbench run directory, or the Tool Meister
+            # sub-system was not orchestrated by pbench-tool-meister-start
+            # for this run.
+            continue
+        yield tm_dir, uuid
+
+
+def gen_host_names(tm_dir: pathlib.Path) -> Iterable[str]:
+    """Read the registered tool data saved for this run and return the list
+    of remote hosts.
+    """
+    run_dir = tm_dir.parent
+    tool_groups = list(gen_tool_groups(run_dir))
+    if not tool_groups:
+        return
+
+    lrh = LocalRemoteHost()
+
+    for tg in tool_groups:
+        for host_name in tg.hostnames.keys():
+            if lrh.is_local(host_name):
+                continue
+            yield host_name
+
+
+class KillTools(BaseCommand):
+    """Find and stop all orchestrated Tool Meister instances."""
+
+    def execute(self, uuids: List[str]) -> int:
+        """Execute the tools kill operation.
+
+        If any UUIDs are passed as arguments, we only want to look for, and
+        locally kill, processes having those UUIDs.
+
+        Without command line arguments, kill all the local PIDs from all the
+        discovered run directories.
+
+        All the Redis server PIDs are killed first, then the Tool Data Sinks,
+        and finally the local Tool Meisters.
+
+        We then remotely kill (via `ssh`) all the Tool Meisters by invoking
```

Comment:
That is the explicit goal.  Kill the Redis servers, that "should" cause all the TMs and the TDS to exit on their own, gracefully.  Anything left over really needs to be killed.

Reply:
We are not hoping they terminate on their own.  We are killing the Redis server first so that they "can't" continue on their own, and to nudge them along to exiting while we work on the remotes.

I fail to see why this order is bad and worth changing.